#### Problem 1

Two independent measurements, of two different pieces of wood, gives the results
$x=95.8\pm 0.1$ cm
and
$y=2.30\pm 0.02$ cm.

>a) Determine the diffence in length, $l=x-y$.

>b) Determine the relative uncertainty in the difference in length, $l$.

<details> <summary>Hint a) click here</summary> Do you add uncertaintites or add them in quadrature? </details>

In [9]:
from uncertainties import ufloat

x = ufloat(95.8, 0.1)
y = ufloat(2.30, 0.02)

l = x - y

print(f"l = {l}")

relative_uncertainty = l.std_dev / l.nominal_value
print(f"Relative uncertainty = {relative_uncertainty:.4e} "
      f"(≈ {relative_uncertainty*100:.2f} %)")

l = 93.50+/-0.10
Relative uncertainty = 1.0907e-03 (≈ 0.11 %)


#### Problem 2

To determine the velocity of a cart on an air track we measure the distance travelled, $d$, and the time interval, $t$.
The measurement values are $d=5.10\pm 0.01$ m, and $t=6.02\pm 0.02$ s.

>a) Determine the velocity of the cart.

>b) Determine the momentum, $p=m\cdot v$, for the cart which has mass $0.711\pm 0.002$ kg.

<details> <summary>Hint a) click here</summary> Do you consider absolute or relative uncertainties here? </details>

In [9]:
from uncertainties import ufloat

d = ufloat(5.10, 0.01)
t = ufloat(6.02, 0.02)

v = d / t

print(f"a)  v = {v}")


m = ufloat(0.711, 0.002)
p = m * v

print(f"b)  p = {p}")


a)  v = 0.8472+/-0.0033
b)  p = 0.6023+/-0.0029


#### Problem 3

A projectile motion starts with initial speed, $v_0=15.3\pm 0.1 \text{m/s}$, and initial angle, $30\pm 1 \degree$.
Assume that the acceleration of gravity, $g=9.82 \text{m/s}^2$, has no uncertainty.

> a) What is the range of the projectile?

> b) What is the range if the uncertainty in the initial speed is $0.2 \text{m/s}$?

> c) What is the relative contribution from uncertainties in initial speed and initial angle to the uncertainty in the range of the projectile?

<details> <summary>Hint a) click here</summary> Is there a formula for the range of a projectile? </details>
<details> <summary>Hint b) click here</summary> Can you make a small change without having to redo the calculations form a)? </details>
<details> <summary>Hint c) click here</summary> ufloat have a method error_components() which may be useful </details>

In [14]:
from uncertainties import ufloat
from uncertainties.umath import sin

g = 9.82  # m/s^2

def projectile_range(v0, theta_deg):
    """v0: ufloat[m/s], theta_deg: ufloat[deg]"""
    theta_rad = theta_deg * (3.141592653589793 / 180.0)
    return v0**2 * sin(2 * theta_rad) / g

# part (a)
v0_a = ufloat(15.3, 0.1)
theta = ufloat(30.0, 1.0)

R_a = projectile_range(v0_a, theta)

print("Part (a):")
print("Range R =", R_a)
print("Nominal R =", R_a.n, "m")
print("Uncertainty dR =", R_a.s, "m")
print()

# part (b)
v0_b = ufloat(15.3, 0.2)

R_b = projectile_range(v0_b, theta)

print("Part (b):")
print("Range R =", R_b)
print("Nominal R =", R_b.n, "m")
print("Uncertainty dR =", R_b.s, "m")
print()

# part (c)
print("Part (c): variance contributions for case (a)")
components_a = R_a.error_components()

for var, contrib in components_a.items():
    print(f"  From {var}: dR_component = {contrib:.6f} m")


var_total_a = R_a.s**2
for var, contrib in components_a.items():
    frac = (contrib**2) / var_total_a
    print(f"  From {var}: fraction of variance = {frac:.3f}")

print()

print("Part (c): variance contributions for case (b)")
components_b = R_b.error_components()

var_total_b = R_b.s**2
for var, contrib in components_b.items():
    frac = (contrib**2) / var_total_b
    print(f"  From {var}: fraction of variance = {frac:.3f}")


Part (a):
Range R = 20.6+/-0.5
Nominal R = 20.644387654979557 m
Uncertainty dR = 0.49590853388936473 m

Part (b):
Range R = 20.6+/-0.7
Nominal R = 20.644387654979557 m
Uncertainty dR = 0.6814694415046798 m

Part (c): variance contributions for case (a)
  From 30.0+/-1.0: dR_component = 0.416053 m
  From 15.30+/-0.10: dR_component = 0.269861 m
  From 30.0+/-1.0: fraction of variance = 0.704
  From 15.30+/-0.10: fraction of variance = 0.296

Part (c): variance contributions for case (b)
  From 30.0+/-1.0: fraction of variance = 0.373
  From 15.30+/-0.20: fraction of variance = 0.627


/var/folders/q1/6y3ztkfn3y3cqm2xmmtytr700000gn/T/ipykernel_10212/3951269693.py:36: FutureWarning: AffineScalarFunc.error_components() is currently an instance method. This method is deprecated. In a future release it will be replaced with an instance property by the same name. It will be accessed by AffineScalarFunc.error_components (with no parentheses).
  components_a = R_a.error_components()
/var/folders/q1/6y3ztkfn3y3cqm2xmmtytr700000gn/T/ipykernel_10212/3951269693.py:50: FutureWarning: AffineScalarFunc.error_components() is currently an instance method. This method is deprecated. In a future release it will be replaced with an instance property by the same name. It will be accessed by AffineScalarFunc.error_components (with no parentheses).
  components_b = R_b.error_components()


#### Problem 4

A car travels in a circular path with radius $R=45\pm 1$ m. 
The car starts from rest, and moves with a tangential acceleration, $a_\text{tan}=0.23\pm 0.01 \text{m/s}^2$.

>a) Determine the magnitude of the acceleration of the car at time $t=10.0\pm 0.5$ s.

>b) Which quantity has the largest contribution to the uncertainty in the acceleration?

>c) Does any quantity have no contribution to the uncertainty in the acceleration? Hint: Consider the properly rounded result.

<details> <summary>Hint a) click here</summary> Remember that the acceleration in circular motion has two different components; there are formulas for both. </details>
<details> <summary>Formula hint a) click here</summary> a(rad) = v**2/R, a(tan)=d|v|/dt </details>
<details> <summary>Hint b+c) click here</summary> You can use the method error_components(). You may wish to label your ufloats, i.e. ufloat(2.0,0.1,'x').  </details>

#### Problem 5

A box slides down an inclined plane. The inclined plane makes an angle of $\theta=30\pm 1 \degree$ with horizontal.
The coefficient of friction between the box and the inclined plane is $\mu_\text{k}=0.22\pm 0.01$.
Assume that the acceleration of gravity, $g=9.82 \text{m/s}^2$ is exact.


>a) Determine the acceleration of the box.

<details> <summary> Formula hint a) click here</summary> a = g * (sin(theta) - mu * cos(theta)) </details>

#### Problem 6

A free fall experiment with a basket ball is used to determine the value of the acceleration of gravity. The determined values for each experiment is found below in the array *g_values*.

>a) Determine the value of $g$ with uncertainty?

>b) Compare the found value of $g$ with uncertainty to the accepted value of $g=9.82 \text{ m/s}^2$

>c) Determine the speed (including uncertainty) of the basket ball if it is dropped a height $1.00\pm 0.02$ m, and you use the largest value from the experiment with uncertainty.

<details> <summary>Hint a) click here</summary> Remember to use the mean and the standard deviation of the mean. </details>
<details> <summary>Hint b) click here</summary> The ufloat method standard_score(value) could be used here. </details>
<details> <summary>Hint c) click here</summary> Remember to use the standard deviation as uncertainty, not the standard deviation of the mean. </details>

In [4]:
import numpy as np
import uncertainties as unc
g_values = np.array([9.71,9.84,9.81,9.73,9.68,9.74])

#### Problem 7

The mechanical energy of a system consisting of a spring and a mass is given by

$E=\frac{1}{2}mv^2+\frac{1}{2}kx^2$.

Assume that all quanties have uncertainties: $m\pm\delta m$, $v\pm\delta v$, $k\pm\delta k$, and $x\pm\delta x$.

>a) Derive a general formula for the uncertainty in the mechanical energy, $\delta E$.

>b) Determine the mechanical energy with uncertainty when $m=0.20\pm 0.01$ kg, $v=5.24\pm 0.02$ m/s, $k=30.0\pm\ 0.5$ N/m, and $x=0.231\pm 0.005$ m.

<details> <summary>Hint a) click here</summary> Use the formula for error propagation. First create a sympy variable with the expression for the energy. </details>
<details> <summary>Hint b) click here</summary> You may use the method subs({variable:value}). </details>